apps/scratch/src/hooks/useEmbeddedScratch.ts

```diff
@@ -33,12 +33,78 @@ export const useEmbeddedScratch = (
           break;
         case "getSubmission":
           if (vm) {
-            respondToMessageEvent({
-              procedure: "getSubmission",
-              result: new Blob([vm.toJSON()], {
-                type: "application/json",
-              }),
-            });
+            let assertionsEnabled = false;
+            const setAssertionsEnabled = (enabled: boolean): void => {
+              assertionsEnabled = enabled;
+            };
+
+            vm.runtime.once(
+              "ARE_ASSERTIONS_ENABLED_RESPONSE",
+              setAssertionsEnabled,
+            );
+
+            // check if assertions are enabled
+            vm.runtime.emit("ARE_ASSERTIONS_ENABLED_QUERY");
+
+            if (!assertionsEnabled) {
+              respondToMessageEvent({
+                procedure: "getSubmission",
+                result: {
+                  file: new Blob([vm.toJSON()], {
+                    type: "application/json",
+                  }),
+                  totalTests: 0,
+                  passedTests: 0,
+                },
+              });
+
+              return;
+            }
+            // if assertions are enabled, we need to run the tests before submitting
+
+            // first stop the project and reset the state
+            vm.runtime.stopAll();
+
+            // then backup project state
+            Promise.all([
```

Comment:
since everything is async from here onwards, can we put it into an async function instead of chaining and nesting `then/catch` callbacks ? I'd be fine keeping the final `catch` inline, to make it explicit that everything's happening asynchronously hereafter.